README.md
![Version 1.1.0](http://img.shields.io/badge/version-v1.1.0-purple.svg)
![Python 3.8](http://img.shields.io/badge/python-3.8-yellow.svg)
[![GPL License](http://img.shields.io/badge/license-GPL%20License-blue.svg)](https://github.com/cinerieus/as3nt/blob/master/LICENSE)  

## As3nt
Another Subdomain ENumeration Tool - written in python to enumerate and enrich subdomains using passive OSINT.  

As3nt can target TLDs or subdomains. The enumeration uses; VirusTotal, HackerTarget, ThreatMiner, urlscan.io and crt.sh. Each subdomain IP is resolved using public DNS servers and the data is enriched using ipwhois and Shodan. As3nt currently outputs to terminal or csv. 

Along with using the tags from Shodan, this tool also adds it's own tags. Currently there are tags added based on certain HTML content found by Shodan like tomcat, fortinet, netscaler and pulse vpn. Also if any CVEs are found for an IP/Subdomain with a CVSS score >= 7.8, these assets are tagged to highlight a possible exploit. 

## Screenshots 
![As3nt_1](https://github.com/cinerieus/as3nt/blob/master/screenshots/as3nt_1.gif)  
![As3nt_2](https://github.com/cinerieus/as3nt/blob/master/screenshots/as3nt_2.gif)

## Installation 
1. Install:
  - with pip: `pip install as3nt`
  - from git (source): 
	- `git clone https://github.com/cinerieus/as3nt.git && cd as3nt/`  
	- `pip install -r requirements.txt`
	- `python ./as3nt/core.py` 
  - from git (releases):
	- `wget https://github.com/cinerieus/as3nt/archive/v1.1.0.tar.gz`
	- `pip install v1.1.0.tar.gz` 
2. Remember to check PATH if you installed in ~/.local/bin/ 
3. Profit! 

*For Shodan functionality set the environment variable 'SHODANKEY' with your API key. 

## Dependencies 
See [requirements.txt](https://github.com/cinerieus/as3nt/blob/master/requirements.txt) 

## Changelog
See [CHANGELOG.md](https://github.com/cinerieus/as3nt/blob/master/CHANGELOG.md) 

## Usage 
![usage](https://github.com/cinerieus/as3nt/blob/master/screenshots/usage.png)  
*If selected Shodan is rate-limited to 1 IP per second.

#### Examples:  
- Run all modules against 'example.com' and save results to csv:  
`as3nt -t example.com -11 -o results.csv`   
- Run against a subdomain:  
`as3nt -s -t subdomain.example.com -11 -o results.csv`  

## Thanks  
- Thanks to [aboul3la](https://github.com/aboul3la/) for the inspiration from [sublist3r](https://github.com/aboul3la/Sublist3r)


CHANGELOG.md
# Changelog 

## 1.1.0
- Removed BufferOver & ThreatCrowd feeds as they are no longer available.

## 1.0.9
- Fixed issue with bulk ASN lookups.

## 1.0.8
- Added CSV header code.

## 1.0.7
- Improved reults handling for less errors when writing data to file.

## 1.0.6
- Fixed broken dependencies.

## 1.0.5
- Fixed deprecated dns resolver config.

## 1.0.4
- Added check for data returned from Virustotal feed.
- Added a few things to help with clean exiting.

## 1.0.3
- Added more error handling for crt.sh source.
- Added fix for different format from BufferOver.

## 1.0.2 
- Removed stray full stop in subenum module. 

## 1.0.1
- Updated error messages. 
- Changed string for pulse vpn tag to stop false positives. 
- Added handling for no subdomains found. 
- Removed unused option. 
- Lowered criteria for CVSS score being tagged for possible exploit. 
- Repo cleanup and started git releases.


as3nt/subenum.py
#####################################################
# Name: subenum
# Author: cinerieus
# Description: Subdomain enumeration module for As3nt 
#####################################################

import sys
import json
import requests
from termcolor import colored
from concurrent.futures import ThreadPoolExecutor

class SubEnum:
    # class init, defines class variables
    def __init__(self,tld):
        self.sublist = []
        self.tld = tld
        # sources defined here, they can be excluded by commenting out
        self.sources = [
                self.VirusTotal,
                self.HackerTarget, 
                self.ThreatMiner,
                self.urlscan_io,
                self.crt_sh
                ]

    def main(self):
        print(colored('Starting subdomain enumeration for '+self.tld+'...', 'magenta'))
        
        # define thread pool with maximum of 20 threads
        with ThreadPoolExecutor(max_workers=20) as pool:
            # tasks is defined here to allow thread tracking for progress check
            tasks = {}
            # loops over source list and submits a task to the thread pool for each source
            for s in self.sources:
                source = s.__name__
                print(colored('[+] '+source, 'cyan'))
                tasks[source] = (pool.submit(s))
            # submits a task to the thread pool for each existsing task, this is used to track the progress or the enumeration
            pool.map(self.progress, tasks.items())

        result = self.inscope(list(set(self.sublist)))
        print(colored('\nTotal: '+str(len(result)), 'yellow'))
       
        # run as module check, allows being run standalone
        if __name__ == '__main__':
            fn = self.tld.replace('.','')+'.txt'
            with open(fn, 'w') as f:
                for r in sorted(result):
                    f.write(r+'\n')
            print(colored('Results saved to: '+fn, 'green'))
        else:
            return sorted(result)

    # progress check for enum progress
    def progress(self,tasks):
        while tasks[1].done() == False:
            pass
        print(colored('[-] '+tasks[0]+' done.', 'green'))
    
    # check if returned subdomains are in the scope of the target
    def inscope(self,sublist):
        for i,v in enumerate(sublist):
            if self.tld not in v:
                del sublist[i]
        return sublist
    
    # VirusTotal feed
    def VirusTotal(self):
        try:
            urls = ['https://www.virustotal.com/ui/domains/'+self.tld+'/subdomains?limit=40']
            for url in urls:
                get = requests.get(url)
                jsondata = json.loads(get.text)
                if 'data' in jsondata.keys():
                    for x in jsondata['data']:
                        self.sublist.append(x['id'])
                    if 'next' in jsondata['links']:
                        urls.append(jsondata['links']['next'])
        except KeyboardInterrupt:
            raise
        except:
            print(colored('[-] VirusTotal - Hit rate limiting!', 'red'))
    
    # HackerTarget feed
    def HackerTarget(self):
        try:
            get = requests.get('https://api.hackertarget.com/hostsearch/?q='+self.tld)
            if get.text == 'API count exceeded - Increase Quota with Membership':
                print(colored('[-] HackerTarget - API quota exceeded!', 'red'))       
            elif get.text != 'error check your search parameter':
                result = get.text.split('\n')
                for r in result:
                    self.sublist.append(r.split(',')[0])
        except KeyboardInterrupt:
            raise
        except:
            print(colored('[-] HackerTarget - Error in response!', 'red'))       

    # ThreatMiner feed
    def ThreatMiner(self):
        try:
            get = requests.get('https://api.threatminer.org/v2/domain.php?q='+self.tld+'&rt=5')
            result = json.loads(get.text)
            if result['status_code'] != '404':
                self.sublist.extend(result['results'])
        except KeyboardInterrupt:
            raise
        except:
            print(colored('[-] ThreatMiner - Error in response!', 'red'))       
    
    # urlscan.io feed
    def urlscan_io(self):
        try:
            get = requests.get('https://urlscan.io/api/v1/search/?q=domain:'+self.tld)
            result = json.loads(get.text)
            if result['total'] != '0':
                for r in result['results']:
                    self.sublist.append(r['page']['domain'])
        except KeyboardInterrupt:            
            raise
        except:
            print(colored('[-] urlscan.io - Error in response!', 'red'))

    # crt.sh feed
    def crt_sh(self):
        try:
            get = requests.get('https://crt.sh/?q=.'+self.tld+'&output=json')
            if get.status_code == 200:
                result = json.loads(get.text)
                if len(result) != 0:
                    for r in result:
                        if '\n' in r['name_value']:
                            subs = r['name_value'].split('\n')
                        else:
                            subs = [r['name_value']]
                        for s in subs:
                            if '*' not in s:
                                self.sublist.append(s)
        except KeyboardInterrupt:
            raise
        except:
            print(colored('[-] crt.sh - Error in response!', 'red'))

# check if being used as module, allows being run standalone
if __name__ == '__main__':
    subenum = SubEnum(sys.argv[1])
    subenum.main()


as3nt/core.py
#########################################
# Name: as3nt
# Author: cinerieus
# Description: Main script for As3nt
##########################################

import os
import sys
import csv
import time
import hashlib
import ipwhois
import argparse
import subprocess
import dns.resolver
from tqdm import tqdm
from shodan import Shodan
from datetime import datetime
from termcolor import colored
from elasticsearch import Elasticsearch
from concurrent.futures import ThreadPoolExecutor

try:
    from as3nt.subenum import SubEnum
except:
    from subenum import SubEnum

class As3nt:
    #class initialisation, declares instance variables and calls main()
    def __init__(self,target,threads,asn,shodan,output,shodankey,subdomains,subonly,bb):
        self.datadict = {}
        self.target = target
        self.threads = threads
        self.asn = asn
        self.shodan = shodan
        self.output = output
        self.shodankey = shodankey
        self.subdomains = subdomains
        self.subonly = subonly
        self.bb = bb

    def getrecords(self,subdomain):
        #handles getting the records for each subdomain, dns server(s) is specified below
        try:
            resolver = dns.resolver.Resolver(configure=False)
            resolver.nameservers = ['1.1.1.1', '1.0.0.1']
            resolver.timeout = 3
            resolver.lifetime = 3
            #gets A records and starts populating main dict, drops CNAME
            try:
                A = resolver.resolve(subdomain, 'A')
                for rec in A.response.answer:
                    if 'CNAME' not in rec.to_text():
                        for x in rec.items:
                            self.datadict[subdomain+x.to_text()] = {'tld':self.target, 'subdomain':subdomain, 'record':'A', 'ip':x.to_text()}
            except:
                pass
            #gets AAAA records, drops CNAME
            try:
                AAAA = resolver.resolve(subdomain, 'AAAA')
                for rec in AAAA.response.answer:
                    if 'CNAME' not in rec.to_text():
                        for x in rec.items:
                            self.datadict[subdomain+x.to_text()] = {'tld':self.target, 'subdomain':subdomain, 'record':'AAAA', 'ip':x.to_text()}
            except:
                pass
            #gets MX records, note these are also a subdomain rather than IP.
            try:
                MX = resolver.resolve(subdomain, 'MX')
                for rec in MX.response.answer:

                    for x in rec.items:
                        self.datadict[subdomain+x.to_text()] = {'tld':self.target, 'subdomain':subdomain, 'record':'MX', 'ip':x.to_text().split(' ')[1]}
            except:
                pass
        except KeyboardInterrupt:
            raise
        except Exception as e:
            print('\nError in getrecords:')
            print(e)
            sys.exit(2)

    def getasn(self,asset):
        #gets ASN data using ipwhois module and RDAP lookup
        try:
            whois = ipwhois.IPWhois(asset['ip'])
            asdata = whois.lookup_rdap(depth=1, retry_count=3, rate_limit_timeout=1, asn_methods=['dns', 'whois', 'http'])
            asn = asdata['asn']
            asndesc = asdata['asn_description']
            try:
                name = asdata['network']['name']
                cidr = asdata['network']['cidr']
            except:
                try:
                    name = asdata['nets'][len(asdata['nets']-1)]['name']
                    cidr = asdata['nets'][len(asdata['nets']-1)]['cidr']
                except:
                    name = 'Not Found'
                    cidr = asdata['adn_cidr']
                    pass
                pass
            self.datadict[asset['subdomain']+asset['ip']].update({'asn':asn, 'asn_description':asndesc, 'asn_netblock':cidr, 'asn_netname':name})
        except KeyboardInterrupt:
            raise
        except Exception as e:
            #print('\nError in getasn:')
            #print(e)
            #sys.exit(2)
            pass

    def getshodan(self,api,asset):
        #get Shodan data per IP, rate limited to 1 per second
        try:
            try:
                # strings to check for in html grabbed by shodan
                taglist = [
                        'tomcat',
                        'fortinet',
                        'netscaler',
                        'pulse connect'
                        ]

                results = api.host(asset['ip'])
                ports = sorted(results['ports'])
                isp = results['isp']
                org = results['org']
                country = results['country_code']
                try:
                    tags = results['tags']
                except:
                    tags = []
                    pass
                try:
                    OS = results['os']
                except:
                    OS = ''
                    pass
                try:
                    vulns = results['vulns']
                except:
                    vulns = []
                    pass

                # loops over shodan data to get html and CVE scores
                for i in range(len(results['data'])):
                    # checks html content for strings is taglist and appends to tags
                    if results['data'][i].get('http'):
                        try:
                            html = results['data'][i]['http']['html'].lower()
                            for t in taglist:
                                if t in html and t not in tags:
                                    tags.append(t)
                        except:
                            pass
                    # checks CVE's for cvss score, if it's "10" tag as possible exploit
                    if results['data'][i].get('vulns'):
                        for r in results['data'][i]['vulns'].values():
                            try:
                                cvss = r['cvss']
                                if cvss >= 7.8:
                                    if 'possible_exploit!' not in tags:
                                        tags.append('possible_exploit!')
                                        break
                            except:
                                pass
            except Exception as e:
                ports = ''
                tags = ''
                isp = ''
                org = ''
                country = ''
                OS = ''
                vulns = ''
                pass
            self.datadict[asset['subdomain']+asset['ip']].update({'shodan_os':OS, 'shodan_tags':tags, 'shodan_ports':ports, 'shodan_vulns':vulns, 'shodan_isp':isp, 'shodan_org':org, 'shodan_country':country})
        except KeyboardInterrupt:
            raise
        except Exception as e:
            #print('\nError in getshodan:')
            #print(e)
            #sys.exit(2)
            pass

    def get_screenshots(self):
        try:
            #subprocess.run(["gowitness", "--disable-db", "file", "-f", "urls.txt", "-P", "./as3nt-elastic/screenshots"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
            p = subprocess.run(["gowitness", "--disable-db", "file", "-f", "urls.txt", "-P", "./as3nt-elastic/screenshots", "-t", str(self.threads)])
            while True:
                if p.stdout:
                    print(p.stdout.next().replace("\n", ""))
                else:
                    break
        except KeyboardInterrupt:
            raise
        except Exception as e:
            print("\nError in get_screenshots")
            print(e)
            #sys.exit(1)
            pass

    def send_data(self, data):
        try:
            data["screenshots"] = []
            sc = os.listdir("./as3nt-elastic/screenshots")
            for p in data["shodan_ports"]:
                fn1 = "http-" + data["subdomain"] + "-" + str(p) + ".png"
                fn2 = "https-" + data["subdomain"] + "-" + str(p) + ".png"
                if fn1 in sc:
                    data["screenshots"].append("http://127.0.0.1:8082/"+fn1)
                if fn2 in sc:
                    data["screenshots"].append("http://127.0.0.1:8082/"+fn2)
            res = self.es.index(index="as3nt-"+self.target+"-"+str(datetime.now().isocalendar()[1])+"-"+str(datetime.now().isocalendar()[0]), id=hashlib.sha1(bytes(data["subdomain"]+data["ip"], 'utf8')).hexdigest(), document=data)
        except KeyboardInterrupt:
            raise
        except Exception as e:
            print("\nError in send_data")
            print(e)
            #sys.exit(1)
            pass

    def run(self):
        try:
            #subdomain input check
            if not self.subdomains:
                subenum = SubEnum(self.target)
                sublist = subenum.main()
                if not sublist:
                    print(colored('No subdomains for: '+self.target+'\n', 'green'))
                    return
                if self.subonly:
                    if self.output:
                        with open(self.output, 'w') as f:
                            for s in sublist:
                                f.write(s+'\n')
                        print(colored('Results saved to: '+self.output+'\n', 'green'))
                    return
            else:
                sublist = self.target
                self.target = 'N/A'
            #threading for i/o heavy tasks, fetches dns records and asn data for each asset
            with ThreadPoolExecutor(max_workers=int(self.threads)) as pool:
                print(colored('\nGetting DNS records...', 'magenta'))
                list(tqdm(pool.map(self.getrecords, sublist), total=len(sublist)))
                if self.asn:
                    print(colored('\nGetting ASN data...', 'magenta'))
                    aslist = list(self.datadict.values())
                    list(tqdm(pool.map(self.getasn, aslist), total=len(aslist)))
            #shodan option check, limited to 1 ip per second by api
            if self.shodan:
                api = Shodan(self.shodankey)
                print(colored('\nGetting Shodan data...', 'magenta'))
                for asset in tqdm(self.datadict.values()):
                    self.getshodan(api,asset)
                    time.sleep(1)
            dictlist = list(self.datadict.values())
            #output option check, writes to csv also checks for existing file to prevent duplicating header
            if self.bb:
                docs = []
                urls = []
                self.es = Elasticsearch(maxsize=self.threads)
                for item in dictlist:
                    if "shodan_ports" in item:
                        if len(item["shodan_ports"]) != 0:
                            docs.append(item)
                            for p in item["shodan_ports"]:
                                urls.append(item["subdomain"]+":"+str(p))

                with ThreadPoolExecutor(max_workers=int(self.threads)) as pool:
                    print(colored('\nSending to elastic...', 'magenta'))
                    list(tqdm(pool.map(self.send_data, docs), total=len(docs)))

                print(colored('\nGetting screenshots...', 'magenta'))
                with open("urls.txt", "w") as f:
                    for url in urls:
                        f.write(url+"\n")
                self.get_screenshots()
                os.remove("urls.txt")
                print(colored('Screenshots saved to: ./as3nt-elastic/screenshots', 'green'))

                with ThreadPoolExecutor(max_workers=int(self.threads)) as pool:
                    print(colored('\nUpdating elastic...', 'magenta'))
                    list(tqdm(pool.map(self.send_data, docs), total=len(docs)))
                print(colored('Bounty hunt complete!', 'green'))

            elif self.output:
                keylist = [list(x.keys()) for x in dictlist]
                header = max(keylist, key=len)
                if not os.path.isfile(self.output):
                    with open(self.output, 'w') as f:
                        w = csv.DictWriter(f, header, extrasaction='ignore')
                        w.writeheader()
                        w.writerows(dictlist)
                    print(colored('Results saved to: '+self.output, 'green'))
                else:
                    with open(self.output, 'a') as f:
                        w = csv.DictWriter(f, header, extrasaction='ignore')
                        w.writerows(dictlist)
                    print(colored('Results saved to: '+self.output, 'green'))
            else:
                for item in dictlist:
                    print(item)
        except KeyboardInterrupt:
            raise
        except Exception as e:
            print('\nError in run:')
            print(e)
            sys.exit(2)

def main():
    # argument declaration
    parser = argparse.ArgumentParser(description='Another Subdomain ENumeration Tool', usage='as3nt -t example.com -11 -o results.csv')
    parser._action_groups.pop()
    required = parser.add_argument_group('Required arguments (-t or -f)')
    optional = parser.add_argument_group('Optional arguments')
    required.add_argument('-t', action='store', dest='target', help='A target TLD or subdomain.')
    required.add_argument('-f', action='store', dest='targetfile', help='A target file that contains a list of TLDs or subdomains.')
    required.add_argument('-o', action='store', dest='output', help='Outputs to a csv.')
    optional.add_argument('-s', action='store_true', dest='subdomains', help='Use for inputing a list of subdomains.')
    optional.add_argument('-so', action='store_true', dest='subonly', help='If specified, will only perform subdomain enumeration.')
    optional.add_argument('-td', action='store', dest='threads', help='Specify number of threads used (defaults to 40).', default=40)
    optional.add_argument('-11', action='store_true', dest='eleven', help='Choose this option to enable all modules.')
    optional.add_argument('-as', action='store_true', dest='asn', help='This option enables the ASN data module.')
    optional.add_argument('-sh', action='store_true', dest='shodan', help='This option enables the Shodan data module.')
    optional.add_argument('-bb', action='store_true', dest='bb', help='This option enables screenshots and es data ingest.')
    args = parser.parse_args()

    # banner
    print(colored("""
                      ____        _
            /\       |___ \      | |
           /  \   ___  __) |_ __ | |_
          / /\ \ / __||__ <| '_ \| __|
         / ____ \\\__ \___) | | | | |_
        /_/    \_\___/____/|_| |_|\__|
    """, 'magenta'))
    print(colored("""
            Written by - @cinereus
    """, 'yellow'))
    print(colored("""
       Another Subdomain ENumeration Tool
    """, 'cyan'))

    # checks for empty/incompatible args
    if not args.target and not args.targetfile or args.target and args.targetfile:
        #parser.print_help()
        print(colored("""
usage: as3nt -t example.com -11 -o results.csv

Required arguments (-t or -f):
-t TARGET      A target TLD or subdomain.
-f TARGETFILE  A target file that contains a list of TLDs or subdomains.
-o OUTPUT      Outputs to a csv.

Optional arguments:
-s             Use for inputing a list of subdomains.
-so            If specified, will only perform subdomain enumeration.
-td THREADS    Specify number of threads used (defaults to 40).
-11            Choose this option to enable all modules.
-as            This option enables the ASN data module.
-sh            This option enables the Shodan data module.
        """, 'green'))
        print('\n')
        sys.exit(0)

    #check for 11 option
    if args.eleven:
        args.asn = True
        args.shodan = True

    # format input
    if args.target:
        target = [args.target]
    else:
        target = []
        with open(args.targetfile, 'r') as csvin:
            reader = csv.reader(csvin)
            rawlist = list(reader)
            for x in rawlist:
                target.append(x[0])

    #shodan api key check
    try:
        shodankey = os.environ['SHODANKEY']
        if shodankey == '':
            print('\nShodan API key not set, disabling Shodan module...\n')
            args.shodan = False
    except:
        print('\nShodan API key not set, disabling Shodan module...\n')
        shodankey = ''
        args.shodan = False
        pass

    #checks for list of subdomains or tlds
    if not args.subdomains:
        try:
            for t in target:
                as3nt =  As3nt(t,args.threads,args.asn,args.shodan,args.output,shodankey,args.subdomains,args.subonly,args.bb)
                as3nt.run()
                print('\n')
        except KeyboardInterrupt:
            print(colored('Exiting...', 'red'))
            exit(0)
        except Exception as e:
            print('\nError in main:')
            print(e)
            sys.exit(1)
    else:
        try:
            as3nt = As3nt(target,args.threads,args.asn,args.shodan,args.output,shodankey,args.subdomains,args.subonly,args.bb)
            as3nt.run()
        except KeyboardInterrupt:
            print(colored('Exiting...', 'red'))
            exit(0)
        except Exception as e:
            print('\nError in main:')
            print(e)
            sys.exit(1)

if __name__ == '__main__':
    main()
